swarm/network/pss_test.go

```diff
@@ -118,86 +540,399 @@ func TestPssTwoRelaySelf(t *testing.T) {
+		id:          nid,
+		network:     net,
+		trigger:     trigger,
+		ct:          ct,
 	}
-	net.SetNaf(naf)
 
 	srv := func(p Peer) error {
-		p.Register(&PssMsg{}, ps.HandlePssMsg)
-		pp.Add(p)
+		p.Register(&PssMsg{}, handlefunc)
+		node.Add(p)
 		p.DisconnectHook(func(err error) {
-			pp.Remove(p)
+			hive.Remove(p)
 		})
 		return nil
 	}
-	protocall := Bzz(addr.OverlayAddr(), addr.UnderlayAddr(), ct, srv, nil, nil).Run
 
-	s := p2ptest.NewProtocolTester(t, NodeId(addr), n, protocall)
+	node.run = Bzz(addr.OverlayAddr(), addr.UnderlayAddr(), ct, srv, nil, nil).Run
+	nodeAdapter.Run = node.run
+
+	return node
+}
+
+func newPssProtocolTester(t *testing.T, ps *Pss, addr *peerAddr, numsimnodes int, handlefunc func(interface{}) error) (*p2ptest.ProtocolTester, *protocols.CodeMap) {
+	testnode := newPssTester(t, ps, addr, numsimnodes, handlefunc, nil, nil)
+	/*protocall := func(p *p2p.Peer, rw p2p.MsgReadWriter) error {
+		glog.V(logger.Detail).Infof("in protocall peer %v rw %v", p, rw)
+		discreason := p.Run()
+		return fmt.Errorf("%v", discreason.String())
+	}*/
+	//ptt := p2ptest.NewProtocolTester(t, NodeId(addr), numsimnodes, protocall)
+	ptt := p2ptest.NewProtocolTester(t, NodeId(addr), numsimnodes, testnode.ProtoCall())
+	return ptt, testnode.ct
+}
+
+func makePss(addr []byte, cachettl time.Duration) *Pss {
+	kp := NewKadParams()
+	kp.MinProxBinSize = 3
+
+	overlay := NewKademlia(addr, kp)
+	ps := NewPss(overlay, cachettl)
+	return ps
+}
+
+func makeCustomProtocol(name string, version int, ct *protocols.CodeMap, id *adapters.NodeId, testpeer *PssTestPeer) *p2p.Protocol {
+	run := func(p *protocols.Peer) error {
+		glog.V(logger.Detail).Infof("running vprotocol: %v", p)
+		/*ptp := &PssTestPeer{ // analogous to bzzPeer in the Bzz() protocol constructor
+			Peer: p,
+			successC: make(chan bool),
+			hasProtocol: true,
+		}*/
+		if testpeer != nil {
+			testpeer.Peer = p
+		}
+		//testpeer.hasProtocol = true
+		
+		//(*testpeers)[id] = ptp
+		//glog.V(logger.Detail).Infof("added testpeer: %v id %v to map %x", (*testpeers)[id], id, testpeers)
+		p.Register(&PssTestPayload{}, testpeer.SimpleHandlePssPayload)
+		err := p.Run()
+		return err
+	}
+
+	return protocols.NewProtocol(name, uint(version), run, ct, nil, nil)
+}
+
+// does exactly what it says
+func makeFakeMsg(ps *Pss, ct *protocols.CodeMap, topic PssTopic, senderaddr PeerAddr, content string) PssMsg {
+	data := PssTestPayload{}
+	code, found := ct.GetCode(&data)
+	if !found {
+		return PssMsg{}
+	}
+
+	data.Data = content
+
+	rlpbundle, err := MakeMsg(code, data)
+	if err != nil {
+		return PssMsg{}
+	}
+
+	pssenv := PssEnvelope{
+		SenderOAddr: senderaddr.OverlayAddr(),
+		SenderUAddr: senderaddr.UnderlayAddr(),
+		Topic:       topic,
+		TTL:         DefaultTTL,
+		Payload:     rlpbundle,
+	}
+	pssmsg := PssMsg{
+		To:      ps.Overlay.GetAddr().OverlayAddr(),
+		Payload: pssenv,
+	}
+
+	return pssmsg
+}
+
+func makePssHandleForward(ps *Pss) func(msg interface{}) error {
```

Comment:
See f.ex line 797, the handler for a PssMsg is not defined from within Pss. This specific example is just a shallow handler for test purposes.